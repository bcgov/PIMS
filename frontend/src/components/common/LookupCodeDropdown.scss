@import '../../colors.scss';
button.map-filter-dropdown {
  background-color: #fff;
  min-width: 7rem;
  color: $dropdown-background-color;
  border: 1px solid $light-variant-color;
  transition: none;
  opacity: 1;
  font-size: 14px;
  font-family: 'BCSans', Fallback, sans-serif;
  padding: 0.5rem;
}

.dropdown {
  .dropdown-menu {
    margin-top: 0;
  }
  .show {
    max-height: 25rem;
    overflow-y: auto;
  }
}
.error.dropdown {
  button {
    border: 1px solid $danger-color !important;
    color: $danger-color;
  }
}
