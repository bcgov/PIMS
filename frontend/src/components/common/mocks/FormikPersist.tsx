import React from 'react';

const Persist = () => <div data-testid="mocked-persist" />;

export { Persist };
