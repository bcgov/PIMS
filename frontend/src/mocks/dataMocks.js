export const createMockHeader = () => ({
  headers: {
    "Access-Control-Allow-Origin": "*",
    Authorization: "mockToken",
  },
});

export const ERROR = { message: "Errors", status: 400 };
