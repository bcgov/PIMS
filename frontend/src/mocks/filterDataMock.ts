import { ILookupCode } from '../actions/lookupActions';
import { IParcel, IProperty } from 'actions/parcelsActions';
import { MapViewportChangeEvent } from 'components/maps/leaflet/Map';

export const SELECTEDCLASSIFICATION = {
  name: 'Core Operational',
  id: 0,
  isDisabled: false,
  type: 'Classification',
};

export const DISABLED = [
  {
    name: 'Core Operational',
    id: 0,
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Core Strategic',
    id: 1,
    isDisabled: true,
    type: 'Classification',
  },
];

export const CLASSIFICATIONS = [
  {
    name: 'Core Operational',
    id: '0',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Core Strategic',
    id: '1',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Surplus Active',
    id: '2',
    isDisabled: false,
    type: 'Classification',
  },
  {
    name: 'Surplus Encumbered',
    id: '3',
    isDisabled: false,
    type: 'Classification',
  },
] as ILookupCode[];

export const AGENCIES = [
  {
    code: 'AEST',
    name: 'AEST',
    id: '1',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'HTLH',
    name: 'HTLH',
    id: '2',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'MOTI',
    name: 'MOTI',
    id: '3',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'FLNR',
    name: 'FLNR',
    id: '4',
    isDisabled: false,
    type: 'Agency',
  },
  {
    code: 'MAH',
    name: 'MAH',
    id: '5',
    isDisabled: false,
    type: 'Agency',
  },
];

export const PARCELS = [
  { id: 1, latitude: 48, longitude: 123 },
  { id: 2, latitude: 50, longitude: 133 },
] as IProperty[];

export const mockDetails = [
  {
    id: 1,
    pid: '000-000-000',
    pin: '',
    projectNumber: '',
    statusId: undefined,
    municipality: '',
    zoning: '',
    zoningPotential: '',
    classificationId: undefined,
    agencyId: undefined,
    isSensitive: false,
    latitude: 48,
    longitude: 123,
    propertyStatus: 'active',
    classification: 'Core Operational',
    description: 'test',
    evaluations: [
      {
        date: new Date(),
        key: 'Assessed',
        value: 10000,
      },
    ],
    fiscals: [
      {
        fiscalYear: 2020,
        key: 'NetBook',
        value: 10000,
      },
    ],
    address: {
      line1: '1234 mock Street',
      line2: 'N/A',
      city: 'Victoria',
      province: 'BC',
      postal: 'V1V1V1',
    },
    landArea: 'unknown',
    landLegalDescription: 'test',
    buildings: [],
    agency: 'MOTI',
  },
  {
    id: 2,
    pid: '000-000-000',
    pin: '',
    statusId: undefined,
    municipality: '',
    zoning: '',
    zoningPotential: '',
    classificationId: undefined,
    agencyId: undefined,
    isSensitive: false,
    latitude: 50,
    longitude: 133,
    propertyStatus: 'active',
    classification: 'Core Operational',
    description: 'test',
    address: {
      line1: '1234 mock Street',
      line2: 'N/A',
      city: 'Victoria',
      province: 'BC',
      postal: 'V1V1V1',
    },
    landArea: 'unknown',
    landLegalDescription: 'test',
    buildings: [],
    evaluations: [
      {
        date: new Date(),
        key: 'Assessed',
        value: 10000,
      },
    ],
    fiscals: [
      {
        fiscalYear: 2020,
        key: 'NetBook',
        value: 10000,
      },
    ],
    agency: 'HLTH',
  },
] as IParcel[];

export const ACTIVE = {
  id: 1,
  pid: '000-000-000',
  pin: '',
  projectNumber: '',
  statusId: undefined,
  municipality: '',
  zoning: '',
  zoningPotential: '',
  classificationId: undefined,
  agencyId: undefined,
  isSensitive: false,
  latitude: 48,
  longitude: 123,
  propertyStatus: 'active',
  classification: 'Core Operational',
  description: 'test',
  evaluations: [
    {
      date: new Date(),
      key: 'Assessed',
      value: 10000,
    },
  ],
  fiscals: [
    {
      fiscalYear: 2020,
      key: 'NetBook',
      value: 10000,
    },
  ],
  address: {
    line1: '1234 mock Street',
    line2: 'N/A',
    city: 'Victoria',
    province: 'BC',
    postal: 'V1V1V1',
  },
  landArea: 'unknown',
  landLegalDescription: 'test',
  buildings: [],
  agency: 'FIN',
} as IParcel;

export const mockAgencyModel = {
  bounds: null,
  filter: {
    agencies: '1',
    classificationId: null,
  },
} as MapViewportChangeEvent;
