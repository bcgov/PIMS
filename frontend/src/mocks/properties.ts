import {
  IAddress,
  IBuilding,
  IEvaluation,
  IFiscal,
  IParcel,
  IProperty,
} from 'actions/parcelsActions';
import { PropertyTypes } from 'constants/index';

export const mockProperty: IProperty = {
  id: 1,
  propertyTypeId: PropertyTypes.PARCEL,
  agencyId: 1,
  agency: 'agency',
  subAgency: 'subagency',
  agencyFullName: 'agency full name',
  subAgencyFullName: 'subagency full name',
  latitude: 0.31,
  longitude: 1.34,
  name: 'name',
  description: 'description',
  projectNumbers: ['SPP-1233'],
  projectStatus: 'status',
  projectWorkflow: 'workflow',
  isSensitive: false,
  createdOn: Date.UTC.toString(),
  updatedOn: Date.UTC.toString(),
  updatedByEmail: 'john.doe@email.com',
  updatedByName: 'John Doe',
};

export const mockAddress: IAddress = {
  line1: '1234 Some Street',
  administrativeArea: 'administrative area',
  provinceId: 'BC',
  postal: 'V9V 3I3',
};

export const mockEvaluation: IEvaluation = {
  key: 'Assessed',
  value: 1000.23,
};

export const mockFiscal: IFiscal = {
  key: 'Appraised',
  value: 333.21,
};

export const mockBuilding: IBuilding = {
  ...mockProperty,
  parcelId: 1,
  pid: 12333,
  address: mockAddress,
  buildingConstructionTypeId: 1,
  buildingPredominateUseId: 1,
  buildingOccupantTypeId: 1,
  classificationId: 1,
  classification: 'classification',
  encumbranceReason: 'encumbranceReason',
  occupantName: 'occupantName',
  transferLeaseOnSale: false,
  buildingTenancy: 'buildingTenancy',
  rentableArea: 234,
  totalArea: 500,
  agencyCode: 'agencyCode',
  assessedLand: 3434.23,
  assessedBuilding: 34343.34,
  evaluations: [mockEvaluation],
  fiscals: [mockFiscal],
  parcels: [],
};

export const mockParcel: IParcel = {
  ...mockProperty,
  pid: '133-123-123',
  pin: '',
  classification: 'classification',
  classificationId: 1,
  encumbranceReason: 'encumbranceReason',
  address: mockAddress,
  landArea: 235,
  landLegalDescription: 'landLegalDescription',
  zoning: 'zoning',
  zoningPotential: 'zoningPotential',
  buildings: [mockBuilding],
  parcels: [],
  assessedLand: 12313.12,
  assessedBuilding: 2432.23,
  evaluations: [mockEvaluation],
  fiscals: [mockFiscal],
  rowVersion: 'version',
};
