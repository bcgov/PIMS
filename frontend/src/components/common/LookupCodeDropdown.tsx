import React, { useState } from 'react';
import { Dropdown, DropdownButton } from 'react-bootstrap';
import _ from 'lodash';
import { ILookupCode } from 'actions/lookupActions';
import { FormikProps } from 'formik';
import './LookupCodeDropdown.scss';

type LookupCodeDropdownProps = {
  defaultTitle: string;
  lookupCodes: ILookupCode[];
  onSelectCode: (codeId: string | null) => void;
  name?: string;
  hideAny?: boolean;
  errors?: boolean;
};

const getNameById = (codeSet: ILookupCode[], codeId: string) => {
  return _.find(codeSet, ['id', codeId])?.name || '';
};

/**
 * Generic component used to display a dropdown list of {@link ILookupCode}
 * @param props {@link LookupCodeDropdownProps}
 */
function LookupCodeDropdown(props: LookupCodeDropdownProps) {
  const [codeDropdownTitle, setCodeDropdownTitle] = useState<string>(props.defaultTitle);

  const onCodeItemChange = (eventKey: string) => {
    if (parseInt(eventKey) === -1) {
      setCodeDropdownTitle('Any');
      props.onSelectCode(null);
      return;
    }
    const agencyName = getNameById(props.lookupCodes, eventKey);
    setCodeDropdownTitle(agencyName);
    props.onSelectCode(eventKey);
  };

  return (
    <DropdownButton
      id="dropdown-basic-button"
      title={codeDropdownTitle}
      bsPrefix="map-filter-dropdown"
      onSelect={onCodeItemChange}
      className={props.errors ? 'error' : null}
    >
      {props.hideAny ? null : <Dropdown.Item eventKey={'-1'}>Any</Dropdown.Item>}
      {props.lookupCodes.map(code => {
        return (
          <Dropdown.Item eventKey={code.id.toString()} key={code.id.toString()}>
            {code.name}
          </Dropdown.Item>
        );
      })}
    </DropdownButton>
  );
}

type FormikLookupCodeDropdownProps = {
  defaultTitle: string;
  lookupCodes: ILookupCode[];
  name: string;
};
export const FormikLookupCodeDropdown = <T extends unknown>(
  props: FormikLookupCodeDropdownProps & FormikProps<T>,
) => {
  if (!props.name) {
    throw new Error('Formik field name must be provided');
  }
  const onSelectCode = (codeId: string | null): void => {
    const nameVal: string = String(props.name);
    props.setFieldValue(nameVal, codeId);
  };
  const isError = () => {
    return _.has(props.errors, props.name) && _.has(props.touched, props.name);
  };
  return (
    <LookupCodeDropdown
      errors={isError()}
      hideAny={true}
      onSelectCode={onSelectCode}
      defaultTitle={props.defaultTitle}
      lookupCodes={props.lookupCodes}
    ></LookupCodeDropdown>
  );
};

export default LookupCodeDropdown;
