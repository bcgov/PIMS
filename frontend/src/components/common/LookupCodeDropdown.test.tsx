import * as MOCK from 'mocks/filterDataMock';
import LookupCodeDropdown from './LookupCodeDropdown';
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme from 'enzyme';
import { Dropdown, DropdownButton } from 'react-bootstrap';

Enzyme.configure({ adapter: new Adapter() });

it('LookupCodeDropdown renders correctly', () => {
  const tree = renderer
    .create(
      <LookupCodeDropdown
        lookupCodes={MOCK.CLASSIFICATIONS}
        defaultTitle={'View by Classification \u00A0'}
        onSelectCode={jest.fn()}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Check items being displayed in dropdown before filter tests
let classificationDropdown = shallow(
  <LookupCodeDropdown
    lookupCodes={MOCK.CLASSIFICATIONS}
    defaultTitle={'View by Classification \u00A0'}
    onSelectCode={jest.fn()}
  />,
);
it('Displays item for all classifications given', () => {
  const items = classificationDropdown.find(Dropdown.Item);
  expect(items.length).toBe(5);
});

let agencyDropdown = shallow(
  <LookupCodeDropdown
    lookupCodes={MOCK.AGENCIES}
    defaultTitle={'View by Agency \u00A0'}
    onSelectCode={jest.fn()}
  />,
);
it('Displays item for all agencies given', () => {
  const items = agencyDropdown.find(Dropdown.Item);
  expect(items.length).toBe(6);
});

// Should isDisabled items be hidden from dropdown? Will fail in current implementation
// it('Does not display item in dropdown list if it is disabled', () => {
//     let disabledDropdown = shallow(<LookupCodeDropdown lookupCodes={MOCK.DISABLED} defaultTitle={"View by Classification \u00A0"} onSelectCode={jest.fn()} />)
//     expect(disabledDropdown.find(Dropdown.Item).length).toBe(2);
// });

// Ensure that when item is selected appropriate function is called
it('Calls onSelect function when item is selected', () => {
  const onSelectAgencies = jest.fn();
  let dropdown = shallow(
    <LookupCodeDropdown
      lookupCodes={MOCK.AGENCIES}
      defaultTitle={'View by Agency \u00A0'}
      onSelectCode={onSelectAgencies}
    />,
  );
  const dropdownButton = dropdown.find(DropdownButton);
  dropdownButton.simulate('select', { eventKey: '1' });
  expect(onSelectAgencies).toBeCalledTimes(1);
});
