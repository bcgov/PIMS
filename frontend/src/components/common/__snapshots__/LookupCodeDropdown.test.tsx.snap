// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LookupCodeDropdown renders correctly 1`] = `
<div
  className="dropdown"
  onKeyDown={[Function]}
>
  <button
    aria-expanded={false}
    aria-haspopup={true}
    className="dropdown-toggle map-filter-dropdown map-filter-dropdown-primary"
    disabled={false}
    id="dropdown-basic-button"
    onClick={[Function]}
    type="button"
  >
    View by Classification Â 
  </button>
</div>
`;
